File: pages/refactorings/replace-inheritance-with-natural-transformation/03-either-option/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

-- Numerals around zero hit both error branches; letter noise
-- covers the unparsable, digit runs the plain positives.
genRaw :: Gen String
genRaw = Gen.choice
  [ show <$> Gen.int (Range.linear (-20) 20)
  , Gen.string (Range.linear 0 4) Gen.alpha
  , Gen.string (Range.linear 1 4) Gen.digit
  ]

prop_quantity_agrees :: Property
prop_quantity_agrees = property $ do
  raw <- forAll genRaw
  Before.quantity raw === After.quantity raw

prop_toMaybe_natural :: Property
prop_toMaybe_natural = property $ do
  raw <- forAll genRaw
  let e = After.parse raw
  After.toMaybe ((* 2) <$> e) === ((* 2) <$> After.toMaybe e)

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("quantity: Before == After", prop_quantity_agrees)
    , ("toMaybe is natural in a", prop_toMaybe_natural)
    ]
  unless ok exitFailure

File: pages/refactorings/replace-inheritance-with-natural-transformation/03-either-option/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("quantity: Before == After", quantityAgrees),
    property("toOption is natural in A", toOptionIsNatural),
  )

  // Numerals around zero hit both error branches; letter noise
  // covers the unparsable, digit runs the plain positives.
  val genRaw: Gen[String] =
    Gen.choice1(
      Gen.int(Range.linear(-20, 20)).map(_.toString),
      Gen.string(Gen.alpha, Range.linear(0, 4)),
      Gen.string(Gen.digit, Range.linear(1, 4)),
    )

  def quantityAgrees: Property =
    for raw <- genRaw.forAll
    yield Before.quantity(raw) ==== After.quantity(raw)

  def toOptionIsNatural: Property =
    for raw <- genRaw.forAll
    yield
      val e = After.parse(raw)
      After.toOption(e.map(_ * 2)) ====
        After.toOption(e).map(_ * 2)

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(
      t.withConfig(PropertyConfig.default), t.result, Seed.fromTime())
    println(
      Test.renderReport("Props", t, r, ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)
